### ISSUE TITLE
Generator Examples/Templates

### ISSUE THREAD

**Issue by PhilippSalvisberg** [oddgen/oddgen#16]
The example generators along with real generators should be provided with an oddgen for SQL Developer release. The installation scripts are part of the extension and may be loaded into a SQL Worksheet for installation in a chosen database schema. For that purpose an additional top level node "Provided Generators" should be introduced in the navigator tree.

This is not a replacement for a public oddgen repository but it's a good starting point. The generators should be managed in a dedicated structure of this github repository.

**Comment by MikeKutz** [comment]
I smell "feature creep".  Here is why...

Server Side generators won't be able to do anything with information on the Client's machine.

For small, MadLib style templates, having the Client send the template code to the server-side generator will be suffice.

However, if you are using a server side generator and a series internal templates (eg a "large template" that is constructed by using a series of smaller, purpose built templates as opposed to a single monolithic template)

You will need to:
- GET the templates from the internal oddgen plugin (or other location such as the client's filesystem or GitHub)
- PUT the templates someplace that the server can access.
- and list the available templates.  RFC-4918 calls this interface PROPFIND

I call the "collection of templates that are stored _place a location name here_" a "Template Repository".
(see associated workgroup for more information)

I recommend you schedule a web meeting to discuss interface(s) and possible features.

**Comment by PhilippSalvisberg** [comment]
As mentioned, it's not a replacement for the template/generator repository as discussed in https://community.oracle.com/docs/DOC-920429 . This feature allows to install generators (e.g. examples) which are part of the oddgen extension (phyiscally part of the .zip file). This feature will not allow to install other generators, so there won't be a public interface.

Please discuss template/generator requirements in the OTN forum, since it is not related with this enhancement.

**Comment by PhilippSalvisberg** [comment]
Changed the title. The idea is not only to "paste" an existing example into a new worksheet, but also a "template" generator which may be used as a starting point for a new database server generator. Instead of adding a new node in the navigator tree a context menu on the the "Database Server Generators" node might be better.

**Comment by PhilippSalvisberg** [comment]
## Implementation Specification

Implement this feature using provided client generators. Representation in navigator tree:
- All Generators
  - Client Generators
    - oddgen Templates
      - PL/SQL
      - Xtend
    - oddgen Examples
      - 1:1 View
      - Hello World
  - Database Server Generators
    - ...

This are two client generators. One named "oddgen Templates, the other named "oddgen Examples". There is one sublevel (object_type) only. There are no further child nodes (object_name). 

This design is dependent on #28. 

### oddgen Templates

This generator has the following parameters:
- Template name (PL/SQL/Xtend)
- Generate files? (boolean, default no)
- Output directory (root folder for the output directory)
- Package name (enabled for PL/SQL and Xtend, default depending on template name)
- Class name (enabled for Xtend generator only)
- Display name (for get_name)
- Description (for get_description)
- Object types (comma separated list of object types for get_object_types)
- Parameters (comma separated list of parameters, default "P1, P2, Is P3?")
- List-of-Values? (boolean, default yes, generates example list-of-values for each parameter, Parameters ending on "?" will get a lead to a Yes-No valid list)

This parameters may be handled by the standard generate dialog. However, to support a directory chooser, the code has to be extended. The idea is to use conventions to identify Parameters with a directory chooser. Parameters ending on "directory" will get a directory chooser in the generate dialog unless a list-of-value is defined for this parameter.

The generation process works as for all other generators. The creation of files is the responsiblity of the client generator. E.g. for Xtend it will produce a full Maven directory structure with a pom.xml and a Xtend file. The content of the POM and Xtend file will be also returned by the generator and depending on the invokation type presented in a new worksheet or copied to the clipboard.

### oddgen Examples

This generator has the following parameters:
- Template language (PL/SQL, FTLDB, tePLSQL, Xtend)
- Package name (default is depending on template language)
- Class name (enabled for Xtend only)

The generation process works as for all other generators. Based on the chosen template language the templates will be read from resource files within the oddgen for SQL Developer JAR file and processed accordingly. The templates for this generator must not contain duplicate existing code. Hence it's the easiest to use the original examples and do some classic search/reaplace using regular expressions for the few parameter replacements for package name and class name).

**Comment by peter-genesys** [comment]
Hi Phillipp,

My name is Peter Burgess.  I've recently joined the ODDGEN project.  

I have a question about the implimentation specification.  I hope this in an appropriate place to ask this question.  Please redirect me if it is not.

Does it currently support generation that does not relate to any specific object or object type in the database?  Can the interface support this without substantial redesign.

I ask because i have a use case, where i use generators to export data, eg config data, into scripts that can be version controlled externally, and executed on database as needed, eg for a release.

These scripts are not simple insert statements, but a set of procedure calls to packages such as TAPIs.  I've attached a set of such generators that i created for QCGU.

[QCGU_generators.zip](https://github.com/oddgen/oddgen/files/331039/QCGU_generators.zip)

As these generators do not relate to any object, but include their own data queries, they would need to be launchable from the Generators view in SQLdev without finding any related objects.

**Simple example**

[DEFTABARRAY]sec_roles[WHERE]role_type like '{role_type_filter}'[ORDERBY]role_type
[FOREACH]sec_roles
[set]destname[to]role_[role_type/u]
prompt Role Config for [role_type/u]
prompt Generated with QCGU
prompt using script "gen_sec_roles.sql"
prompt GUID [script_universal_id/e]

begin

  sec_utils.add_sec_roles(i_role_type    => '[role_type/u]'
                         ,i_role_name    => '[role_name]'
                         ,i_role_desc    => '[role_desc/2]'
                         ,i_active_code  => '[active_code/u]'
                         ,i_auto_inherit => '[auto_inherit/u]');

end;
/

commit;

[ENDFOREACH]

Thanks

Peter.

**Comment by PhilippSalvisberg** [comment]
> I have a question about the implimentation specification. I hope this in an appropriate place to ask this question.

Yes, if it is about the about the "Implementation Specification" I've written above. For other topics please [submit a new issue](https://github.com/oddgen/oddgen/issues/new).

> Does it currently support generation that does not relate to any specific object or object type in the database?

Yes, even with the current version v0.2, since you may give object_type and object_name a own meaning. See the dropall generator example where an object type "All" is used. However issue #28 deals with a more flexible structure in the navigator tree. In version v0.2 you have at least to define a dummy object_type and an dummy object_name. In v0.3 this will not be necessary anymore.
